src/MoBi.Presentation/Presenter/EditFormulaInPathAndValuesPresenter.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System.Linq;
+using MoBi.Core.Commands;
+using MoBi.Core.Domain.Model;
+using MoBi.Core.Domain.Services;
+using MoBi.Core.Helper;
+using MoBi.Presentation.Mappers;
+using MoBi.Presentation.Tasks.Interaction;
+using MoBi.Presentation.Views;
+using OSPSuite.Core.Domain;
+using OSPSuite.Core.Domain.Builder;
+using OSPSuite.Core.Services;
+using OSPSuite.Presentation.Presenters;
+
+namespace MoBi.Presentation.Presenter
+{
+   public interface IEditFormulaInPathAndValuesPresenter : ICommandCollectorPresenter, ISubjectPresenter, IEditFormulaPresenter
+   {
+      void Init<TBuilder>(TBuilder formulaOwner, IBuildingBlock buildingBlock, UsingFormulaDecoder formulaDecoder, MoBiMacroCommand macroCommand) where TBuilder : PathAndValueEntity, IUsingFormula, IWithDisplayUnit;
+   }
+
+   public class EditFormulaInPathAndValuesPresenter : EditFormulaPresenter<IEditFormulaInPathAndValuesView, IEditFormulaInPathAndValuesPresenter>, IEditFormulaInPathAndValuesPresenter
+   {
+      private readonly IInteractionTaskContext _interactionTaskContext;
+      private MoBiMacroCommand _localMacroCommand;
+
+      public EditFormulaInPathAndValuesPresenter(IEditFormulaInPathAndValuesView view, 
+         IFormulaPresenterCache formulaPresenterCache, 
+         IMoBiContext context,
+         IFormulaToFormulaInfoDTOMapper formulaDTOMapper, 
+         FormulaTypeCaptionRepository formulaTypeCaptionRepository,
+         IMoBiFormulaTask formulaTask, 
+         ICircularReferenceChecker circularReferenceChecker,
+         IInteractionTaskContext interactionTaskContext,
+         ISelectReferenceAtParameterValuePresenter referencePresenter) : base(view, formulaPresenterCache, context, formulaDTOMapper, formulaTask, formulaTypeCaptionRepository, circularReferenceChecker)
+      {
+         _interactionTaskContext = interactionTaskContext;
+         ReferencePresenter = referencePresenter;
+      }
+
+      public void FormulaTypeSelectionChanged(string formulaName)
+      {
+         rollBackChanges();
+
+         var (command, newFormula) = _formulaTask.CreateNewFormulaInBuildingBlock(_formulaDTO.Type, FormulaDimension, AllFormulaNames, _buildingBlock, formulaName);
+         if (newFormula == null)
+            return;
+
+         AddCommand(command);
+         SelectFormula(newFormula);
+         UpdateFormula();
+      }
+
+      private void rollBackChanges()
+      {
+         if (_localMacroCommand.IsEmpty) 
+            return;
+
+         _interactionTaskContext.CancelCommand(_localMacroCommand);
```

Comment:
I Wonder how well this works...feels dangerous to me lol. Can't we just swap out the formula?

Reply:
Let me see how this can be minimally refactored for it. The base presenter uses the formulaOwner and buildingBlock for some stuff. Maybe we can get around it easily

Reply:
All the underlying specific formula type editors perform their actions using commands, implying that the intention is to operate directly on the formula while it's in-place.

Rather than refactor all that, I have done this by cloning the building block and operating on the cloned building block formula cache and the cloned formula owner.

Then those commands are just thrown away and we use an update formula command and an add to cache command to make the change.

Reply:
yeah good idea. 
I just did not expect the undo to work as smooth as we expected it...

Reply:
It did work, but I was uneasy about it. Since you also had reservations, I revised it.